src/cliques.c

```diff
@@ -0,0 +1,628 @@
+////////////////////////////////////////////////////////////////////////////////
+//
+// cliques.c               di/cliques                        Julius Jonusas
+//
+// Copyright (C) 2020 - Julius Jonusas 
+//
+// This file is free software, see the digraphs/LICENSE.
+
+
+// C headers
+#include <setjmp.h>   // for longjmp, setjmp, jmp_buf
+#include <stdbool.h>  // for true, false, bool
+#include <stddef.h>   // for NULL
+#include <stdint.h>   // for uint16_t, uint64_t
+#include <stdlib.h>   // for malloc, NULL
+
+// GAP headers
+#include "src/compiled.h"
+
+// Digraphs package headers
+#include "bitarray.h"         // for BitArray
+#include "conditions.h"       // for Conditions
+#include "digraphs-debug.h"   // for DIGRAPHS_ASSERT
+#include "homos-graphs.h"     // for Digraph, Graph, . . .
+#include "perms.h"            // for MAXVERTS, UNDEFINED, PermColl, Perm
+
+////////////////////////////////////////////////////////////////////////////////
+// Macros
+////////////////////////////////////////////////////////////////////////////////
+
+#define MIN(a, b) (a < b ? a : b)
+
+////////////////////////////////////////////////////////////////////////////////
+// Forward declarations
+////////////////////////////////////////////////////////////////////////////////
+
+// Defined in digraphs.h
+Int DigraphNrVertices(Obj);
+Obj FuncOutNeighbours(Obj, Obj);
+Obj FuncADJACENCY_MATRIX(Obj, Obj);
+
+// GAP level things, imported in digraphs.c
+extern Obj IsDigraph;
+extern Obj Infinity;
+extern Obj IsSymmetricDigraph;
+extern Obj GeneratorsOfGroup;
+extern Obj AutomorphismGroup;
+extern Obj IsPermGroup;
+extern Obj IsDigraphAutomorphism;
+extern Obj LargestMovedPointPerms;
+extern Obj SmallestMovedPointPerm;
+extern Obj IsClique;
+extern Obj IsTrivial;
+extern Obj Orbit;
+extern Obj Stabilizer;
+extern Obj IsSubset;
+extern Obj OnTuples;
+extern Obj Group;
+extern Obj ClosureGroup;
+
+////////////////////////////////////////////////////////////////////////////////
+// Global variables
+////////////////////////////////////////////////////////////////////////////////
+
+static Obj GAP_FUNC;       // Variable to hold a GAP level hook function
+
+static Obj (*HOOK)(void*,  // HOOK function applied to every homo found
+                   const BitArray*,
+                   const uint16_t);
+static void* USER_PARAM;   // A USER_PARAM for the hook
+
+static jmp_buf OUTOFHERE;  // So we can jump out of the deepest
```

Comment:
Just to remark: this code won't work in a multi threaded setting due to all these global variables :-(.

The usual solution I'd use would be to put these into a simple struct, and then pass around a pointer to that struct to all functions that need it. At the top level, a member of that struct can be instantiated on the stack.

This can actually even be beneficial in terms of performance, as on many architectures, accessing global variables is slower than accessing a plain pointer.

Reply:
Thanks for the comments @fingolfin, much of the kernel module of Digraphs is not thread-safe, an issue that I'm aware of, but haven't really seen any reason to change until now.

Reply:
In practice I also don't care that much about thread safety, it is just usually the quickest way to convince people of my true desire: to avoid global variables as much as possible. They tend to cause all kinds of design issues down the road, and often make it much harder to reason about code that uses them. But here at least they are static and restricted to one relatively small source file.

Anyway, all that is just my opinion, and obviously that shouldn't count for much here, as I am not a stakeholder in this project :-). I merely stumbled over this PR and left some well-meant quick comments. Mind you, overall this code looks perfectly fine

Reply:
Thanks @fingolfin, the comments are much appreciated!